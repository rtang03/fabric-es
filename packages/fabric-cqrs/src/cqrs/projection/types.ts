import { Commit } from '../../types';

export interface FindAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      where?: Record<string, any>;
      all?: boolean;
      contain?: string | number;
    };
  };
}

export interface UpsertAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      commit: Record<string, Commit>;
    };
  };
}

export interface UpsertManyAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      commits: Record<string, Commit>;
    };
  };
}
