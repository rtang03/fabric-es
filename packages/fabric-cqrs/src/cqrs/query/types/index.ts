/**
 * @packageDocumentation
 * @hidden
 */
export * from './action';
