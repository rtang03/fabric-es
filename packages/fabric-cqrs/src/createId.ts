export const createId: (parts: string[]) => string = parts => parts[0] + '::' + parts[1];
