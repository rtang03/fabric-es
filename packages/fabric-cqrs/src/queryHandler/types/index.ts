export * from './commitHashfields';
export * from './entityDefaultSchema';
export * from './commitDefaultSchema';
//
export * from './mapField';
export * from './redisCommit';
