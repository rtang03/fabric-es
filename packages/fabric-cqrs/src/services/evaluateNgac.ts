import { Network } from 'fabric-network';
import { getContract } from './contract';

export const evaluateNgac: <T = any>(
  fcn: string,
  args: string[],
  { network }: {network: Network}
) => Promise<T | { error: any }> = async <T = any>(fcn, args, { network }) =>
  await getContract(network).then(
    async ({ contract }) =>
      await contract
        .createTransaction(fcn)
        .evaluate(...args)
        .then<T>((res: any) =>
          JSON.parse(Buffer.from(JSON.parse(res)).toString())
        )
        .catch(error => {
          console.log(`Error processing Evaluate transaction: ${fcn}`);
          // console.error(error.stack);
          return { error };
        })
  );

export default evaluateNgac;
