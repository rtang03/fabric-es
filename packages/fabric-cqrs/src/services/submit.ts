import { from, Observable } from 'rxjs';
import { Commit, Context } from '../types';
import { getContract } from './contract';

export const submit: (
  fcn: string,
  args: string[],
  { network }: Context
) => Promise<
  Record<string, Commit> | { error?: any; status?: string; message?: string }
> = async (fcn, args, { network }) =>
  await getContract(network).then(({ contract }) =>
    contract
      .createTransaction(fcn)
      .submit(...args)
      .then<Record<string, Commit>>((res: any) =>
        JSON.parse(Buffer.from(JSON.parse(res)).toString())
      )
      .catch(error => {
        console.log(`Error processing Submit transaction.`);
        console.error(error.stack);
        return { error };
      })
  );

export const submit$: (
  fcn: string,
  args: string[],
  context: Context
) => Observable<
  Record<string, Commit> | { error?: any; status?: string; message?: string }
> = (fcn, args, context) => from(submit(fcn, args, context));
