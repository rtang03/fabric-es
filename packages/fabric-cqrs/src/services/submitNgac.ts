import { Context } from '../types';
import { getContract } from './contract';

export const submitNgac: <T = any>(
  fcn: string,
  args: string[],
  { network }: Context
) => Promise<T | { error?: any; status?: string; message?: string }> = async <
  T = any
>(
  fcn,
  args,
  { network }
) =>
  await getContract(network).then(({ contract }) =>
    contract
      .createTransaction(fcn)
      .submit(...args)
      .then<T>((res: any) =>
        JSON.parse(Buffer.from(JSON.parse(res)).toString())
      )
      .catch(error => {
        console.log(`Error processing Submit transaction.`);
        console.error(error.stack);
        console.log(fcn);
        return { error };
      })
  );

export default submitNgac;
