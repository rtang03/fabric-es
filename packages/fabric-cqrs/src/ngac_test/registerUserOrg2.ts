import { Context, registerUser as register } from '@espresso/admin-tool';
import { ChannelEventHub } from 'fabric-client';
import { FileSystemWallet, Gateway, Network } from 'fabric-network';
import { getNetwork } from '../services';

const defaultContext: Context = {
  connectionProfile: process.env.CONNECTION_PROFILE,
  fabricNetwork: process.env.NETWORK_LOCATION,
  wallet: new FileSystemWallet(process.env.WALLET)
};

export const registerUserOrg2: (option: {
  enrollmentID: string;
  enrollmentSecret: string;
  context?: Context;
}) => any = async ({
  enrollmentID,
  enrollmentSecret,
  context = defaultContext
}) => register(enrollmentID, enrollmentSecret, context);

export const bootstrap: (
  identity: string
) => Promise<{
  identity: string;
  network: Network;
  gateway: Gateway;
  channelHub: ChannelEventHub;
}> = async identity => {
  try {
    await registerUserOrg2({
      enrollmentID: identity,
      enrollmentSecret: 'password'
    });
    return await getNetwork({ identity });
  } catch (err) {
    console.log(err);
    process.exit(-1);
  }
};
