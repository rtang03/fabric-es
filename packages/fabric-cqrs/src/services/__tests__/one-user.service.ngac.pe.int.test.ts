import { config } from 'dotenv';
import { resolve } from 'path';
config({ path: resolve(__dirname, '../../../.env.ngac.org2.test') });

import { Gateway, Network } from 'fabric-network';
import { omit, pick, values } from 'lodash';
import { createId } from '../../createId';
import { bootstrap } from '../../ngac_test/registerUserOrg2';
import { Attribute, Commit, CONTEXT, RESOURCE } from '../../types';
import evaluateNgac from '../evaluateNgac';
import { submit } from '../submit';
import submitNgac from '../submitNgac';

let network: Network;
let gateway: Gateway;
let fakerNetwork: Network;
let fakerGateway: Gateway;

const identity = `svs_org2_pe_test${Math.floor(Math.random() * 10000)}`;
const faker = `faker_${Math.floor(Math.random() * 10000)}`;

beforeAll(async () => {
  const ctx = await bootstrap(identity);
  network = ctx.network;
  gateway = ctx.gateway;
  const fakerCtx = await bootstrap(faker);
  fakerNetwork = fakerCtx.network;
  fakerGateway = fakerCtx.gateway;
});

afterAll(async () => {
  await gateway.disconnect();
  await fakerGateway.disconnect();
});

const mspId = 'Org2MSP';
const entityName = 'doc';
const entityId = `doc_0_${Math.floor(Math.random() * 10000)}`;
const eventsStr = JSON.stringify([
  { type: 'DocCreated', payload: { data: 'Mydata' } }
]);
const mspAttrs = [{ type: '1', key: 'mspid', value: mspId }];
const pClass = 'ngac-pe-policy';
const sid = 'allowCreateDoc';
const url = `model/${mspId}/${entityName}`;
const condition = JSON.stringify({ hasList: { createDoc: 'creator_id' } });
const allowedEvents = JSON.stringify(['DocCreated']);
const id = createId([mspId, identity]);
const resourceAttr = [{ type: '1', key: 'createDoc', value: id }];

describe('Ngac Permission Tests', () => {
  it('should pass permission check', async () => {
    // create Org2MSP attribute
    await submitNgac('addMSPAttr', [mspId, JSON.stringify(mspAttrs)], {
      network
    }).then(attributes => expect(attributes).toEqual(mspAttrs));

    // should create policy
    await submitNgac(
      'addPolicy',
      [pClass, sid, url, allowedEvents, condition],
      {
        network
      }
    )
      .then(policy => omit(policy, 'key'))
      .then(policy => expect(policy).toMatchSnapshot());

    // create resource attributes for entityName
    await submitNgac(
      'addResourceAttr',
      [entityName, '', JSON.stringify(resourceAttr)],
      { network }
    ).then(attributes => expect(attributes).toEqual(resourceAttr));

    // should createDoc
    await submit('createCommit', [entityName, entityId, '0', eventsStr], {
      network
    })
      .then<Commit>(result => values(result)[0])
      .then(commit => pick(commit, 'entityName', 'version', 'events'))
      .then(commit =>
        expect(commit).toEqual({
          entityName,
          version: 0,
          events: [{ type: 'DocCreated', payload: { data: 'Mydata' } }]
        })
      );
  });
});

describe('Permission Test 2', () => {
  it('should should pass the test, for existing entityId', async () => {
    // create policy 2
    const sid = 'allowUpadateDoc';
    const url = `model/${mspId}/${entityName}/${entityId}`;
    const allowedEvents = JSON.stringify(['DocUpdated']);
    const condition = JSON.stringify({
      hasList: { updateDoc: RESOURCE.CREATOR_ID },
      stringEquals: {
        [CONTEXT.INVOKER_MSPID]: RESOURCE.CREATOR_MSPID,
        [CONTEXT.INVOKER_ID]: RESOURCE.CREATOR_ID
      }
    });
    await submitNgac(
      'addPolicy',
      [pClass, sid, url, allowedEvents, condition],
      {
        network
      }
    )
      .then(policy => omit(policy, 'key', 'attributes'))
      .then(policy => expect(policy).toMatchSnapshot());

    // upsert resourceAttribute of updateDoc
    const attr = JSON.stringify([{ type: 'N', key: 'updateDoc', value: [id] }]);
    await submitNgac('upsertResourceAttr', [entityName, entityId, attr], {
      network
    })
      .then((attributes: Attribute[]) =>
        attributes.filter(({ key }) => key === 'updateDoc')
      )
      .then(attributes => attributes[0])
      .then(({ value }) => expect(value).toEqual([id]));

    // updateDoc
    const eventsStr = JSON.stringify([
      { type: 'DocUpdated', payload: { data: 'mydoc' } }
    ]);
    await submit('createCommit', [entityName, entityId, '1', eventsStr], {
      network
    })
      .then<Commit>(result => values(result)[0])
      .then(commit => pick(commit, 'entityName', 'version', 'events'))
      .then(commit =>
        expect(commit).toEqual({
          entityName,
          version: 1,
          events: [{ type: 'DocUpdated', payload: { data: 'mydoc' } }]
        })
      );
  });
});

describe('Rejected Permission Test', () => {
  it('should fail createNothing, when no policy found', async () =>
    submit(
      'createCommit',
      [
        entityName,
        `doc_1_${Math.floor(Math.random() * 10000)}`,
        '0',
        JSON.stringify([{ type: 'NothingCreated' }])
      ],
      {
        network
      }
    ).then(({ error: { message } }) =>
      expect(message).toContain('No policy found')
    ));
});

describe('NGAC CRUD query', () => {
  it('should getMSPAttrByMSPID', async () =>
    evaluateNgac('getMSPAttrByMSPID', [mspId], { network }).then(attrs =>
      expect(attrs).toEqual([{ type: '1', key: 'mspid', value: 'Org2MSP' }])
    ));

  it('should getResourceAttrByURI', async () => {
    await evaluateNgac(
      'getResourceAttrByURI',
      [`model/Org2MSP/${entityName}`],
      { network }
    ).then(attrs =>
      expect(attrs).toEqual([{ type: '1', key: 'createDoc', value: id }])
    );

    // Below is tested working, but keep comment out, for simplicity
    // await evaluateNgac(
    //   'getResourceAttrByURI',
    //   [`model/Org2MSP/${entityName}/${entityId}`],
    //   { network }
    // ).then(result => console.log(result));

    // Below is tested working, but keep comment out, for simplicity
    // await evaluateNgac('getPolicyById', [id], {
    //   network
    // }).then(result => console.log(result));
  });
});

describe('Permission Test 3', () => {
  it('should fail updateDoc within unauthorized user', async () => {
    const network = fakerNetwork;
    const eventsStr = JSON.stringify([
      { type: 'DocUpdated', payload: { data: 'update me again' } }
    ]);

    await submit('createCommit', [entityName, entityId, '2', eventsStr], {
      network
    }).then(({ error: { message } }) =>
      expect(message).toContain('No policy found')
    );

    // should fail because stringEquals requires that
    // only invoker must be resource creator
    const sid = 'allowUpadateDoc';
    const url = `model/${mspId}/${entityName}/${entityId}`;
    const allowedEvents = JSON.stringify(['DocUpdated']);
    let condition = JSON.stringify({
      // hasList is optional
      hasList: { updateDoc: RESOURCE.CREATOR_ID },
      stringEquals: {
        [CONTEXT.INVOKER_MSPID]: RESOURCE.CREATOR_MSPID,
        // below statement guards the permission
        // if remove, anyone can invoke updateDoc
        // todo: add new function in chaincode for update and upsert policy
        [CONTEXT.INVOKER_ID]: RESOURCE.CREATOR_ID
      }
    });
    await submitNgac(
      'addPolicy',
      [pClass, sid, url, allowedEvents, condition],
      {
        network
      }
    );

    await submit('createCommit', [entityName, entityId, '2', eventsStr], {
      network
    }).then(({ error: { message } }) =>
      expect(message).toContain(`"allowUpadateDoc" assertion fails`)
    );

    // should pass
    condition = JSON.stringify({
      stringEquals: {
        [CONTEXT.INVOKER_MSPID]: RESOURCE.CREATOR_MSPID,
      }
    });
    await submitNgac(
      'addPolicy',
      [pClass, sid, url, allowedEvents, condition],
      {
        network
      }
    );
  });
});
