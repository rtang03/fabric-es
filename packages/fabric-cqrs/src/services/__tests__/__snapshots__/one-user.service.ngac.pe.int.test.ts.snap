// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ngac Permission Tests should pass permission check 1`] = `
Array [
  Object {
    "key": "mspid",
    "type": "1",
    "value": "Org2MSP",
  },
]
`;

exports[`Ngac Permission Tests should pass permission check 2`] = `
Object {
  "allowedEvents": Array [
    "DocCreated",
  ],
  "attributes": Object {
    "uri": "model/Org2MSP/doc",
  },
  "condition": Object {
    "hasList": Object {
      "createDoc": "creator_id",
    },
  },
  "effect": "Allow",
  "policyClass": "ngac-pe-policy",
  "sid": "allowCreateDoc",
}
`;

exports[`Ngac Permission Tests should pass permission check 3`] = `
Array [
  Object {
    "key": "createDoc",
    "type": "1",
    "value": "Org2MSP::Admin@org2.example.com",
  },
]
`;

exports[`Ngac Permission Tests should pass permission check 4`] = `
Object {
  "entityName": "doc",
  "events": Array [
    Object {
      "payload": Object {
        "data": "Mydata",
      },
      "type": "DocCreated",
    },
  ],
  "version": 0,
}
`;

exports[`Permission Test 2 should should pass the test, for existing entityId 1`] = `
Object {
  "allowedEvents": Array [
    "DocUpdated",
  ],
  "attributes": Object {
    "uri": "model/Org2MSP/doc/doc_0_8672",
  },
  "condition": Object {
    "hasList": Object {
      "updateDoc": "creator_id",
    },
    "stringEquals": Object {
      "invoker_mspid": "creator_mspid",
    },
  },
  "effect": "Allow",
  "policyClass": "ngac-pe-policy",
  "sid": "allowUpadateDoc",
}
`;
